File: peras-hs/src/Peras/Message.hs
module Peras.Message where

import Peras.Block (Block, BlockBody, Certificate)
import Peras.Chain (Chain, Vote)
import Peras.Crypto (Hash)
import Peras.Numbering (SlotNumber)

newtype NodeId = MkNodeId {nodeId :: String}
  deriving (Eq, Ord, Read, Show)

data Message
  = NextSlot SlotNumber
  | NewChain Chain
  | SomeVote Vote
  | SomeCertificate Certificate
  | FetchVotes [Hash Vote]
  | FollowChain (Hash Chain)
  | RollForward Block
  | RollBack Block
  | FetchBlocks [Hash Block]
  | SomeBlock BlockBody

File: peras-rust/src/message.rs
use serde::{Deserialize, Serialize};

use crate::block::{Block, BlockBody, Slot};
use crate::chain::{Chain, Vote};
use crate::crypto::Hash;

pub type NodeId = String;

#[derive(Debug, Clone, Deserialize, Serialize, Hash)]
#[serde(tag = "tag", content = "contents")]
pub enum Message {
    // control messages
    Stop,
    NextSlot(Slot),
    // actual protocol messages
    NewChain(Chain),
    SomeVote(Vote),
    FetchVotes(Vec<Hash>),
    FollowChain(Hash),
    RollForward(Block),
    RollBack(Block),
    FetchBlocks(Vec<Hash>),
    SomeBlock(BlockBody),
}

#[derive(Debug, Clone, Deserialize, Serialize)]
#[serde(tag = "tag", content = "contents")]
pub enum OutMessage {
    FetchBlock(Block),
    SendMessage(Message),
}
